anchor_optimization/optimize_anchors.py

```diff
@@ -0,0 +1,347 @@
+
+    overlap = compute_overlap(entries, anchors)
+    max_overlap = np.amax(overlap, axis=1)
+    not_matched = len(np.where(max_overlap < 0.5)[0])
+
+    if mode == 'avg':
+        result = 1 - np.average(max_overlap)
+    elif mode == 'ce':
+        result = np.average(-np.log(max_overlap))
+    elif mode == 'focal':
+        result = np.average(-(1 - max_overlap) ** 2 * np.log(max_overlap))
+    else:
+        raise Exception('Invalid mode.')
+
+    if set_state is not None:
+        state = set_state
+
+    # --------------------------------------------------------------------------------------------------------------------------------
+    # "scipy.optimize.differential_evolution" utilizes multiprocessing but internally uses "multiprocessing.Pool" and not
+    # "multiprocessing.Process" which is required for sharing state between processes
+    # (see: https://docs.python.org/3/library/multiprocessing.html#sharing-state-between-processes)
+    #
+    # the "state" variable does not affect directly the "scipy.optimize.differential_evolution" process, therefore updates will be
+    # printed out in case of improvement only if a single thread is used
+    # --------------------------------------------------------------------------------------------------------------------------------
+
+    if threads == 1:
+
+        if result < state['best_result']:
+            state['best_result'] = result
+
+            if verbose:
+                print('Current best anchor configuration')
+                print('State: {}'.format(np.round(state['best_result'], 5)))
+                print(
+                    'Ratios: {}'.format(
+                        sorted(
+                            np.round(
+                                anchor_params.ratios,
+                                3))))
+                print(
+                    'Scales: {}'.format(
+                        sorted(
+                            np.round(
+                                anchor_params.scales,
+                                3))))
+
+            if include_stride:
+                if verbose:
+                    print(
+                        'Average overlap: {}'.format(
+                            np.round(
+                                np.average(max_overlap),
+                                3)))
+
+            if verbose:
+                print(
+                    "Number of labels that don't have any matching anchor: {}".format(not_matched))
+                print()
+
+    if to_tuple:
+        # return a tuple, which happens in the last call to the 'average_overlap' function
+        return result, not_matched
+    else:
+        return result
+
+
+# ***************************************************************************************************************************************************
+# !!!! the python "anchors_optim" function is meant to be used from the command line (from within a Python console it gives incorrect results)  !!!!
```

Comment:
This is a weird statement, why does it not give correct results? Also can you use docstring comments? https://www.python.org/dev/peps/pep-0257/#id17

Reply:
This is something I can't figure out. Initially I thought that it might had to do with the fact that I had 'keras-retinanet' both in Python2 and Python3 in my OS installed. However, this is not the case because I run the function today in a docker environment where only Python3 is installed and I get different results although using the same seed,

```R

# from the command-line 

State: 0.10112
Ratios: [0.471, 0.731, 1.0, 1.367, 2.123]
Scales: [0.541, 0.655, 0.781, 0.913]
Number of labels that dont have any matching anchor: 2

# from within Python

State: 0.21237
Ratios: [0.449, 0.728, 1.0, 1.373, 2.226]
Scales: [0.4, 0.469, 0.549, 0.656]
Number of labels that dont have any matching anchor: 1653

```
Obviously, from within Python gives the wrong output. I run the same script a couple of days ago also using a single thread and I received similar results.
The code is actually the same as before this PR, just wrapped in a function and exposed to Python (previously this wasn't possible as it was run only from the command line). What do you think?

Reply:
Hmm not sure, I will try to have a look at this but I am a bit short on time so will accept this PR as is.